File: rtdb_bridge/envelope.py
"""
Build the canonical Python -> TS envelope for the dbchart extension host.

The envelope mirrors the TypeScript contracts referenced by the VS Code
extension and is the single object returned by
:meth:`convert_export_to_envelope`:

    {
        "envelopeVersion": 1,
        "generator": { "name": "rtdb-bridge", "version": "0.1.0", "pipelineVersion": 1 },
        "source": { "format": "rtdb-export-json", "name": "...", ... },
        "database": DatabaseSchema,     # canvas contract
        "tree": [SchemaTreeNode],       # explorer tree
        "sql": { "dialect": "postgres", "tables": [...] },
        "warnings": [...]
    }
"""

from __future__ import annotations

import json
from dataclasses import asdict
from typing import Any, Dict, Iterable, List, Optional

from relationalize import Schema
from relationalize.sql_dialects import PostgresDialect

from . import __version__
from .preprocess import DEFAULT_KNOWN_COLLECTION_FIELDS, extract_collection_records
from .pipeline import ExportPipelineResult, TableInfo, relationalize_export
from .schemas import load_schema
from .ui_tree import (
    build_database_tree,
    choice_parts,
    normalize_type_label,
)

# --------------------------------------------------------------------------- #
# Constants
# --------------------------------------------------------------------------- #

ENVELOPE_VERSION: int = 1
PIPELINE_VERSION: int = 1
GENERATOR_NAME: str = "rtdb-bridge"
SOURCE_FORMAT: str = "rtdb-export-json"
DEFAULT_DIALECT: str = "postgres"
DEFAULT_SCHEMA_NAME: str = "public"

_POSTGRES_DIALECT = PostgresDialect()


# --------------------------------------------------------------------------- #
# DatabaseSchema canvas builders
# --------------------------------------------------------------------------- #


def _column_sql_type(raw_type: str) -> str:
    """Map a relationalize raw type to the dialect SQL column type."""
    if raw_type in _POSTGRES_DIALECT.type_column_mapping:
        return _POSTGRES_DIALECT.type_column_mapping[raw_type]
    return "VARCHAR(65535)"


def _build_choice_members(
    raw_rows: List[Dict[str, Any]], clean_schema: Schema, column: str, parts: List[str]
) -> List[Dict[str, Any]]:
    """Build the physical member objects for a logical choice/mixed column."""
    converted = [clean_schema.convert_object(row) for row in raw_rows]
    members = []
    for part in parts:
        if part == "none":
            continue
        phys_name = f"{column}_{part}"
        values = [row.get(phys_name) for row in converted]
        members.append(
            {
                "name": phys_name,
                "type": normalize_type_label(part),
                "sqlType": _column_sql_type(part),
                "nullable": any(v is None for v in values) if values else False,
                "partial": any(phys_name not in row for row in converted),
            }
        )
    return members


def _build_database_table(
    table_info: TableInfo,
    available_tables: set,
) -> Dict[str, Any]:
    """Convert one pipeline table into a DatabaseSchema table object."""
    clean = table_info.clean_schema()
    raw_rows = table_info.rows
    converted = [clean.convert_object(row) for row in raw_rows]

    columns: List[Dict[str, Any]] = []
    for column, raw_type in clean.schema.items():
        parts = choice_parts(raw_type)
        if parts:
            members = _build_choice_members(raw_rows, clean, column, parts)
            columns.append(
                {
                    "name": column,
                    "type": "mixed",
                    "uiType": "mixed",
                    "types": [normalize_type_label(p) for p in parts],
                    "sqlType": None,
                    "nullable": any(m["nullable"] for m in members) if members else False,
                    "partial": any(m["partial"] for m in members) if members else False,
                    "choiceOf": members,
                }
            )
            continue

        values = [row.get(column) for row in converted]
        ui_type = normalize_type_label(raw_type)
        columns.append(
            {
                "name": column,
                "type": ui_type,
                "uiType": ui_type,
                "sqlType": _column_sql_type(raw_type),
                "nullable": (any(v is None for v in values) or any(column not in row for row in converted)) if values else False,
                "partial": any(column not in row for row in converted),
            }
        )

    foreign_keys = []
    for col_obj in columns:
        col_name = col_obj["name"]
        child_name = f"{table_info.name}_{col_name}"
        if child_name in available_tables:
            foreign_keys.append(
                {
                    "name": f"{table_info.name}_{col_name}_fk",
                    "column": col_name,
                    "refTable": child_name,
                    "refColumn": "record_id",
                    "cardinality": "1:N",
                }
            )

    return {
        "name": table_info.name,
        "type": "table",
        "sourceCollection": None,
        "rowCount": len(raw_rows),
        "columnCount": len(columns),
        "columns": columns,
        "primaryKey": ["record_id"],
        "indexes": [],
        "foreignKeys": foreign_keys,
    }


def build_database_schema(
    database_name: str,
    pipeline_result: ExportPipelineResult,
) -> Dict[str, Any]:
    """Build the DatabaseSchema canvas contract from pipeline table outputs."""
    available_tables = set(pipeline_result.tables.keys())
    tables = [
        _build_database_table(info, available_tables)
        for info in pipeline_result.tables.values()
    ]
    tables.sort(key=lambda t: t["name"])
    return {
        "name": database_name,
        "tables": tables,
    }


# --------------------------------------------------------------------------- #
# SQL contract builders
# --------------------------------------------------------------------------- #


def _build_sql_table(table_info: TableInfo) -> Dict[str, Any]:
    """Build one table object within the SQL contract."""
    clean = table_info.clean_schema()
    ddl = clean.generate_ddl(table=table_info.name)

    columns = []
    for column, raw_type in clean.schema.items():
        parts = choice_parts(raw_type)
        if parts:
            sql_type = None
            choice_of = []
            for part in parts:
                if part == "none":
                    continue
                phys_name = f"{column}_{part}"
                choice_of.append(
                    {
                        "type": part,
                        "sqlType": _column_sql_type(part),
                        "columnName": phys_name,
                        "nullable": True,
                        "partial": True,
                    }
                )
            columns.append(
                {
                    "name": column,
                    "rawType": raw_type,
                    "sqlType": sql_type,
                    "choiceOf": choice_of,
                }
            )
        else:
            columns.append(
                {
                    "name": column,
                    "rawType": raw_type,
                    "sqlType": _column_sql_type(raw_type),
                }
            )

    columns.sort(key=lambda c: c["name"])
    return {
        "name": table_info.name,
        "schemaName": DEFAULT_SCHEMA_NAME,
        "ddl": ddl,
        "rowCount": len(table_info.rows),
        "columns": columns,
    }


def build_sql_contract(
    pipeline_result: ExportPipelineResult,
) -> Dict[str, Any]:
    """Build SQL DDL metadata contract from pipeline schema outputs."""
    return {
        "dialect": DEFAULT_DIALECT,
        "tables": [
            _build_sql_table(info)
            for info in sorted(pipeline_result.tables.values(), key=lambda t: t.name)
        ],
    }


# --------------------------------------------------------------------------- #
# Envelope assembly & validation
# --------------------------------------------------------------------------- #


def _build_source_info(
    source_name: Optional[str],
    pipeline_result: ExportPipelineResult,
) -> Dict[str, Any]:
    record_count = sum(c.recordCount for c in pipeline_result.collections)
    return {
        "format": SOURCE_FORMAT,
        "name": source_name,
        "recordCount": record_count,
        "collectionCount": len(pipeline_result.collections),
    }


def build_envelope(
    database_name: str,
    pipeline_result: ExportPipelineResult,
    source_name: Optional[str] = None,
    validate: bool = True,
) -> Dict[str, Any]:
    """
    Assemble the full Python -> TS envelope from pipeline outputs.

    Parameters
    ----------
    database_name:
        Display name for the database/canvas root.
    pipeline_result:
        Output of :meth:`~rtdb_bridge.pipeline.relationalize_export`.
    source_name:
        Optional original database/export name for the ``source`` member.
    validate:
        When True, validate against ``schemas/rtdb-bridge.schema.json``
        using the optional ``jsonschema`` package.  If the package is not
        installed a ``schema-validation-unavailable`` warning is appended and
        no exception is raised.  When False no validation is performed.
    """
    warnings: List[Dict[str, Any]] = []

    envelope: Dict[str, Any] = {
        "envelopeVersion": ENVELOPE_VERSION,
        "generator": {
            "name": GENERATOR_NAME,
            "version": __version__,
            "pipelineVersion": PIPELINE_VERSION,
        },
        "source": _build_source_info(source_name, pipeline_result),
        "database": build_database_schema(database_name, pipeline_result),
        "tree": [build_database_tree(database_name, list(pipeline_result.tables.values()))],
        "sql": build_sql_contract(pipeline_result),
        "warnings": warnings,
    }

    if validate:
        _validate_envelope(envelope, warnings)

    return envelope


def _validate_envelope(
    envelope: Dict[str, Any], warnings: List[Dict[str, Any]]
) -> None:
    """Validate the envelope against the canonical schema when possible."""
    try:
        import jsonschema
        from jsonschema import RefResolver
    except ImportError:  # pragma: no cover - optional dependency
        warnings.append(
            {
                "code": "schema-validation-unavailable",
                "message": (
                    "The 'jsonschema' package is not installed.  The envelope "
                    "was produced but could not be validated against "
                    "rtdb-bridge.schema.json.  Install jsonschema to enable "
                    "contract validation."
                ),
            }
        )
        return

    schema = load_schema("rtdb-bridge")

    # Load every sibling canonical schema so cross-file $refs resolve.
    store = {
        "https://dbchart.dev/contracts/database.schema.json": load_schema("database"),
        "https://dbchart.dev/contracts/schema-tree.schema.json": load_schema("schema-tree"),
        "https://dbchart.dev/contracts/sql.schema.json": load_schema("sql"),
    }
    resolver = RefResolver.from_schema(schema, store=store)

    try:
        jsonschema.validate(envelope, schema, resolver=resolver)
    except jsonschema.ValidationError as exc:
        # Best-effort: surface schema drift as a warning rather than a crash.
        warnings.append(
            {
                "code": "schema-validation-failed",
                "message": f"Envelope failed validation against rtdb-bridge.schema.json: {exc.message}",
                "details": {"path": list(exc.absolute_path), "validator": exc.validator},
            }
        )


def convert_export_to_envelope(
    export: Dict[str, Any],
    database_name: Optional[str] = None,
    source_name: Optional[str] = None,
    known_collection_fields: Optional[Iterable[str]] = None,
    validate: bool = True,
) -> Dict[str, Any]:
    """
    Full bridge: raw RTDB export JSON -> the canonical UI envelope.

    This is the primary entry point for the dbchart extension host and for
    the Python worker.

    Parameters
    ----------
    export:
        Parsed Firebase RTDB export JSON.  Each top-level key is a collection
        name whose value is a dict of Firebase key -> record.
    database_name:
        Display name for the database/canvas root.  Defaults to
        ``"rtdb_to_sql"``.
    source_name:
        Optional original export / database name recorded in ``source``.
        Defaults to ``database_name``.
    known_collection_fields:
        Optional field-name set marking keyed maps collections keyed by
        business ids.  Defaults to
        :data:`~rtdb_bridge.preprocess.DEFAULT_KNOWN_COLLECTION_FIELDS`.
    validate:
        Pass-through to :meth:`build_envelope` controlling optional
        JSON-schema validation (see there for semantics).
    """
    if database_name is None:
        database_name = "rtdb_to_sql"
    if source_name is None:
        source_name = database_name

    pipeline_result = relationalize_export(
        export,
        known_collection_fields=known_collection_fields,
    )
    return build_envelope(
        database_name=database_name,
        pipeline_result=pipeline_result,
        source_name=source_name,
        validate=validate,
    )


__all__ = [
    "ENVELOPE_VERSION",
    "GENERATOR_NAME",
    "PIPELINE_VERSION",
    "SOURCE_FORMAT",
    "build_database_schema",
    "build_envelope",
    "build_sql_contract",
    "convert_export_to_envelope",
]

File: node/src/rtdb_bridge/envelope.ts
/**
 * Build the canonical Python -> TS (now Node -> TS) envelope.
 *
 * Ported from `rtdb_bridge/envelope.py`. The envelope mirrors the TypeScript
 * contracts referenced by the VS Code extension and is the single object
 * returned by {@link convertExportToEnvelope}:
 *
 * ```
 * {
 *   envelopeVersion: 1,
 *   generator: { name: 'rtdb-bridge', version: '0.1.0', pipelineVersion: 1 },
 *   source: { format: 'rtdb-export-json', name: '...', ... },
 *   database: DatabaseSchema,     // canvas contract
 *   tree: [SchemaTreeNode],       // explorer tree
 *   sql: { dialect: 'postgres', tables: [...] },
 *   warnings: [...]
 * }
 * ```
 */
import type { BaseSupportedColumnType } from '../relationalize/types.js';
import type { JsonObject } from '../relationalize/relationalize.js';
import { Schema } from '../relationalize/index.js';
import { PostgresDialect } from '../relationalize/sqlDialects.js';

import {
  GENERATOR_NAME,
  POSTGRES_TYPE_MAPPING,
  choiceParts,
  normalizeTypeLabel,
} from './contracts.js';
import {
  ExportPipelineResult,
  TableInfo,
  relationalizeExport,
} from './pipeline.js';
import { buildDatabaseTree } from './uiTree.js';
import { loadSchema } from './schemas.js';
import {
  validateInstance,
  type SchemaRegistry,
} from './jsonSchema.js';
import { version as packageVersion } from './version.js';

export const ENVELOPE_VERSION = 1;
export const PIPELINE_VERSION = 1;
export { GENERATOR_NAME };
export const SOURCE_FORMAT = 'rtdb-export-json';
export const DEFAULT_DIALECT = 'postgres';
export const DEFAULT_SCHEMA_NAME = 'public';

const POSTGRES_DIALECT = new PostgresDialect();

/** Map a relationalize raw type to the dialect SQL column type. */
function columnSqlType(rawType: string): string {
  return (
    POSTGRES_TYPE_MAPPING[rawType as BaseSupportedColumnType] ??
    'VARCHAR(65535)'
  );
}

/** Build the physical member objects for a logical choice/mixed column. */
function buildChoiceMembers(
  rawRows: ReadonlyArray<JsonObject>,
  cleanSchema: Schema,
  column: string,
  parts: string[],
): Array<Record<string, unknown>> {
  const converted = rawRows.map((row) => cleanSchema.convertObject(row));
  const members: Array<Record<string, unknown>> = [];
  for (const part of parts) {
    if (part === 'none') {
      continue;
    }
    const physName = `${column}_${part}`;
    const values = converted.map((row) => row[physName]);
    members.push({
      name: physName,
      type: normalizeTypeLabel(part),
      sqlType: columnSqlType(part),
      nullable: values.length
        ? values.some((v) => v === null || v === undefined)
        : false,
      partial: converted.some((row) => !(physName in row)),
    });
  }
  return members;
}

/** Convert one pipeline table into a DatabaseSchema table object. */
function buildDatabaseTable(
  tableInfo: TableInfo,
  availableTables: ReadonlySet<string>,
): Record<string, unknown> {
  const clean = tableInfo.cleanSchema();
  const rawRows = tableInfo.rows;
  const converted = rawRows.map((row) => clean.convertObject(row));

  const columns: Array<Record<string, unknown>> = [];
  for (const [column, rawType] of Object.entries(clean.schema)) {
    const parts = choiceParts(rawType);
    if (parts) {
      const members = buildChoiceMembers(rawRows, clean, column, parts);
      columns.push({
        name: column,
        type: 'mixed',
        uiType: 'mixed',
        types: parts.map((p) => normalizeTypeLabel(p)),
        sqlType: null,
        nullable: members.length
          ? members.some((m) => m['nullable'] === true)
          : false,
        partial: members.length
          ? members.some((m) => m['partial'] === true)
          : false,
        choiceOf: members,
      });
      continue;
    }

    const values = converted.map((row) => row[column]);
    const uiType = normalizeTypeLabel(rawType);
    columns.push({
      name: column,
      type: uiType,
      uiType,
      sqlType: columnSqlType(rawType),
      nullable: values.length
        ? values.some((v) => v === null || v === undefined) ||
          converted.some((row) => !(column in row))
        : false,
      partial: converted.some((row) => !(column in row)),
    });
  }

  const foreignKeys: Array<Record<string, unknown>> = [];
  for (const colObj of columns) {
    const colName = colObj['name'] as string;
    const childName = `${tableInfo.name}_${colName}`;
    if (availableTables.has(childName)) {
      foreignKeys.push({
        name: `${tableInfo.name}_${colName}_fk`,
        column: colName,
        refTable: childName,
        refColumn: 'record_id',
        cardinality: '1:N',
      });
    }
  }

  return {
    name: tableInfo.name,
    type: 'table',
    sourceCollection: null,
    rowCount: rawRows.length,
    columnCount: columns.length,
    columns,
    primaryKey: ['record_id'],
    indexes: [],
    foreignKeys,
  };
}

/** Build the DatabaseSchema canvas contract from pipeline table outputs. */
export function buildDatabaseSchema(
  databaseName: string,
  pipelineResult: ExportPipelineResult,
): Record<string, unknown> {
  const availableTables = new Set(pipelineResult.tables.keys());
  const tables = [...pipelineResult.tables.values()].map((info) =>
    buildDatabaseTable(info, availableTables),
  );
  tables.sort((a, b) =>
    (a['name'] as string) < (b['name'] as string)
      ? -1
      : a['name'] === b['name']
        ? 0
        : 1,
  );
  return {
    name: databaseName,
    tables,
  };
}

/** Build one table object within the SQL contract. */
function buildSqlTable(tableInfo: TableInfo): Record<string, unknown> {
  const clean = tableInfo.cleanSchema();
  const ddl = clean.generateDdl(tableInfo.name);

  const columns: Array<Record<string, unknown>> = [];
  for (const [column, rawType] of Object.entries(clean.schema)) {
    const parts = choiceParts(rawType);
    if (parts) {
      const choiceOf: Array<Record<string, unknown>> = [];
      for (const part of parts) {
        if (part === 'none') {
          continue;
        }
        const physName = `${column}_${part}`;
        choiceOf.push({
          type: part,
          sqlType: columnSqlType(part),
          columnName: physName,
          nullable: true,
          partial: true,
        });
      }
      columns.push({
        name: column,
        rawType,
        sqlType: null,
        choiceOf,
      });
    } else {
      columns.push({
        name: column,
        rawType,
        sqlType: columnSqlType(rawType),
      });
    }
  }

  columns.sort((a, b) =>
    (a['name'] as string) < (b['name'] as string)
      ? -1
      : a['name'] === b['name']
        ? 0
        : 1,
  );
  return {
    name: tableInfo.name,
    schemaName: DEFAULT_SCHEMA_NAME,
    ddl,
    rowCount: tableInfo.rows.length,
    columns,
  };
}

/** Build SQL DDL metadata contract from pipeline schema outputs. */
export function buildSqlContract(
  pipelineResult: ExportPipelineResult,
): Record<string, unknown> {
  return {
    dialect: DEFAULT_DIALECT,
    tables: [...pipelineResult.tables.values()]
      .sort((a, b) =>
        a.name < b.name ? -1 : a.name === b.name ? 0 : 1,
      )
      .map((info) => buildSqlTable(info)),
  };
}

function buildSourceInfo(
  sourceName: string | null | undefined,
  pipelineResult: ExportPipelineResult,
): Record<string, unknown> {
  const recordCount = pipelineResult.collections.reduce(
    (sum, c) => sum + c.recordCount,
    0,
  );
  return {
    format: SOURCE_FORMAT,
    name: sourceName ?? null,
    recordCount,
    collectionCount: pipelineResult.collections.length,
  };
}

export interface EnvelopeWarning {
  code: string;
  message: string;
  details?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface RtdbBridgeEnvelope {
  envelopeVersion: number;
  generator: Record<string, unknown>;
  source: Record<string, unknown>;
  database: Record<string, unknown>;
  tree: Array<Record<string, unknown>>;
  sql: Record<string, unknown>;
  warnings: EnvelopeWarning[];
  [key: string]: unknown;
}

function comparePaths(
  a: Array<string | number>,
  b: Array<string | number>,
): number {
  const length = Math.max(a.length, b.length);
  for (let i = 0; i < length; i += 1) {
    const left = a[i];
    const right = b[i];
    if (left === undefined) return -1;
    if (right === undefined) return 1;
    if (left === right) continue;
    return String(left) < String(right) ? -1 : 1;
  }
  return 0;
}

/**
 * Assemble the full envelope from pipeline outputs.
 *
 * When `validate` is true the envelope is validated against
 * `rtdb-bridge.schema.json` (and its cross-file `$ref`s); schema drift is
 * surfaced as a warning rather than a crash, exactly like the Python bridge.
 */
export function buildEnvelope(
  databaseName: string,
  pipelineResult: ExportPipelineResult,
  sourceName?: string | null,
  validate = true,
): RtdbBridgeEnvelope {
  const warnings: EnvelopeWarning[] = [];

  const envelope: RtdbBridgeEnvelope = {
    envelopeVersion: ENVELOPE_VERSION,
    generator: {
      name: GENERATOR_NAME,
      version: packageVersion,
      pipelineVersion: PIPELINE_VERSION,
    },
    source: buildSourceInfo(sourceName, pipelineResult),
    database: buildDatabaseSchema(databaseName, pipelineResult),
    tree: [
      buildDatabaseTree(databaseName, [
        ...pipelineResult.tables.values(),
      ]) as unknown as Record<string, unknown>,
    ],
    sql: buildSqlContract(pipelineResult),
    warnings,
  };

  if (validate) {
    validateEnvelope(envelope, warnings);
  }

  return envelope;
}

/** Validate the envelope against the canonical schema, collecting warnings. */
export function validateEnvelope(
  envelope: RtdbBridgeEnvelope,
  warnings: EnvelopeWarning[],
): void {
  const schema = loadSchema('rtdb-bridge');

  // Register every sibling canonical schema by its absolute $id so
  // cross-file $refs resolve - the Node analogue of the Python `Registry`.
  const registry: SchemaRegistry = new Map();
  registry.set(
    'https://dbchart.dev/contracts/database.schema.json',
    loadSchema('database'),
  );
  registry.set(
    'https://dbchart.dev/contracts/schema-tree.schema.json',
    loadSchema('schema-tree'),
  );
  registry.set(
    'https://dbchart.dev/contracts/sql.schema.json',
    loadSchema('sql'),
  );

  const errors = validateInstance(envelope, schema, registry);
  errors.sort((a, b) => comparePaths(a.path, b.path));

  for (const exc of errors) {
    warnings.push({
      code: 'schema-validation-failed',
      message:
        'Envelope failed validation against rtdb-bridge.schema.json: ' +
        exc.message,
      details: {
        path: exc.path,
        validator: exc.validator,
      },
    });
  }
}

/**
 * Full bridge: raw RTDB export JSON -> the canonical UI envelope.
 *
 * This is the primary entry point for the dbchart extension host and for the
 * Node worker.
 *
 * @param exportTree Parsed Firebase RTDB export JSON. Each top-level key is a
 *   collection name whose value is a dict of Firebase key -> record.
 * @param databaseName Display name for the database/canvas root. Defaults to
 *   `"rtdb_to_sql"`.
 * @param sourceName Optional original export / database name recorded in
 *   `source`. Defaults to `databaseName`.
 * @param knownCollectionFields Optional field-name set marking keyed maps
 *   collections keyed by business ids. Defaults to
 *   `DEFAULT_KNOWN_COLLECTION_FIELDS`.
 * @param validate Controls optional JSON-schema validation of the envelope.
 */
export function convertExportToEnvelope(
  exportTree: JsonObject,
  databaseName?: string | null,
  sourceName?: string | null,
  knownCollectionFields?: Iterable<string> | null,
  validate = true,
): RtdbBridgeEnvelope {
  const resolvedDatabaseName = databaseName ?? 'rtdb_to_sql';
  const resolvedSourceName = sourceName ?? resolvedDatabaseName;

  const pipelineResult = relationalizeExport(
    exportTree,
    knownCollectionFields,
  );
  return buildEnvelope(
    resolvedDatabaseName,
    pipelineResult,
    resolvedSourceName,
    validate,
  );
}
